lib/xctest.js

```diff
@@ -0,0 +1,180 @@
+        const pid = await instrumentService._callChannel(true, instrumentDaemon,
+            'launchSuspendedProcessWithDevicePath:bundleIdentifier:environment:arguments:options:',
+            appPath, this.xctestBundleId, appEnv, appArgs, appOptions);
+        log.info(`Pid of launched ${this.xctestBundleId} ${pid.selector}:`);
+        const msg = new DTXMessageAuxBuffer();
+        msg.appendObject(pid.selector);
+        await instrumentService._callChannel(true, instrumentDaemon, 'startObservingPid:', msg);
+        const logDebug = (message) => {
+            log.debug(`instrument output: ${message.auxiliaries[0]}`);
+        };
+        instrumentService.registerSelectorCallback('_XCT_logDebugMessage', logDebug);
+        return { instrumentService, pid: pid.selector };
+    }
+
+    async start() {
+        if (this.running) {
+            log.warn('already running!');
+            return;
+        }
+        this.running = true;
+        const { DAEMON_CONNECTION_INTERFACE } = TESTMANAGERD_CHANNEL;
+        const productVersion = await getOSVersion(this.udid);
+        const majorVersion = parseInt(productVersion.split('.')[0], 10);
+        const xcodeVersion = 29;
+        const sessionIdentifier = uuidV4();
+        //first connection
+        this._initialControlSession = await startTestmanagerdService(this.udid);
+        if (majorVersion >= 11) {
+            const msg = new DTXMessageAuxBuffer();
+            msg.appendObject(xcodeVersion);
+            await this._initialControlSession.callChannel(DAEMON_CONNECTION_INTERFACE, '_IDE_initiateControlSessionWithProtocolVersion:', msg);
+        }
+
+        //second connection
+        this._execTestPlanSession = await startTestmanagerdService(this.udid);
+        const startExecuting = () => {
+            if (this._executing) {
+                return;
+            }
+            this._executing = true;
+            const msg = new DTXMessageAuxBuffer();
+            msg.appendObject(xcodeVersion);
+            this._execTestPlanSession._callChannel(false, 0xFFFFFFFF, '_IDE_startExecutingTestPlanWithProtocolVersion:', msg);
+        };
+        const showLogMessage = (message) => {
+            log.debug(message.auxiliaries[0]);
+            if (message.auxiliaries.join('').indexOf('Received test runner ready reply with error: (null') !== -1) {
+                log.info('Test runner ready');
+                startExecuting();
+            }
+        };
+        this._execTestPlanSession.registerSelectorCallback(
+            '_XCT_testBundleReadyWithProtocolVersion:minimumVersion:',
+            startExecuting);
+        this._execTestPlanSession.registerSelectorCallback('_XCT_logDebugMessage:', showLogMessage);
+        let msg = new DTXMessageAuxBuffer();
+        msg.appendObject(new NSUUID(sessionIdentifier));
+        msg.appendObject(sessionIdentifier + '-746F-006D726964646C79');
+        msg.appendObject('/Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild');
```

Comment:
Should the result of `xcode-select -p` be better? (or allow to give the `/Applications/Xcode.app/Contents/Developer` as an argument as well? since Appium itself allows users to set an arbitrary path for this part

Reply:
This is just for launching in device. It will work cross-platform, so it doesn't matter what version we are using